### PR TITLE
release/public-v1: mandatory bugfixes and namelist update for MRW App release v1.1

### PR DESCRIPTION
**Pull request by climbfuji**
This PR contains a number of mandatory bugfixes and one update for the UFS MRW App release v1.1. 

Mandatory bugfixes:
- update modulefiles for hera.gnu and jet.intel to reflect changes in modules environment and filesystem structure
- bugfix for jet.intel to pass correct SIMD instruction flags to CCPP (from develop branch)
- update modulefiles for linux.gnu and macosx.gnu so that `compile_cmake.sh` also works with NCEPLIBS-external and NCEPLIBS umbrella build
- update of fv3atm and ccpp-framework submodule pointers for Python 3 bugfix

Update for MRW v1.1:
- parm/ccpp_v16beta_c96.nml.IN: change diffusion according to https://forums.ufscommunity.org/threads/strength-second-order-damping-top-layers; see also https://github.com/ufs-community/ufs-mrweather-app/issues/149#issue-588485247

Associated PRs:
https://github.com/NCAR/ccpp-framework/pull/303
https://github.com/NOAA-EMC/fv3atm/pull/130
https://github.com/ufs-community/ufs-weather-model/pull/149

For regression testing information, see below.

### CODE REVIEW

**Comment by ligiabernardet** [comment]
Shouldn't these innovations go to the v1.1 branch (yet to be created)?

**Comment by climbfuji** [comment]
> Shouldn't these innovations go to the v1.1 branch (yet to be created)?

No, the tag 1.1.0 will be created from this branch (where all the other Python 3 etc updates have gone in already).

**Review by DusanJovic-NOAA — APPROVED** [review]

**Comment by climbfuji** [comment]
Sorry, @DusanJovic-NOAA I accidentally requested a re-review from you. That was unintentional.

**Review by DusanJovic-NOAA — APPROVED** [review]

**Comment by climbfuji** [comment]
Regression testing on hera.intel, cheynne.intel, cheyenne.gnu: first test against existing baseline. All GFS v15p.2 tests pass, all GFS v16beta tests fail (because of the change in diffusion parameters). Then, create new baselines and verify against them: all tests pass. Regression test logs updated in the PR. Ready to merge.

File: parm/ccpp_v16beta_c96.nml.IN [hunk 1]
```diff
@@ -48,8 +48,8 @@ deflate_level=1
   nudge_dz = .false.
   tau = 10.
   rf_cutoff = 7.5e2
-  d2_bg_k1 = 0.20
-  d2_bg_k2 = 0.0
+  d2_bg_k1 = 0.15
+  d2_bg_k2 = 0.02
```
Comment (junwang-noaa) [inline parm/ccpp_v16beta_c96.nml.IN:52]: Do we need to update v16beta namelist? I thought it was fixed at Oct, 2019.

**Review by ligiabernardet — COMMENTED** [review]
This PR is to branch release/public-v1 and the PR discussion says that the branch for the UFS MRW App v1.1 release will be created from this branch. This is inconsistent with my understanding of the plan laid out by Arun in the 6/16 release meeting. I thought a release branch for the v1.1 release would be created from release/public-v1. I am not defending any particular solution, I just think this is inconsistent with my take away from the discussion and needs clarification. Can @arunchawla-NOAA clarify this and/or do we need more discussion?

On the topic of namelist changes, these look okay to me. They were proposed after a conversation with Fanglin as discussed in  https://forums.ufscommunity.org/threads/strength-second-order-damping-top-layers.

File: parm/ccpp_v16beta_c96.nml.IN (hunk 1 above)
Comment (climbfuji) [reply parm/ccpp_v16beta_c96.nml.IN:52]: Please see Ligia's comment below (https://github.com/ufs-community/ufs-weather-model/pull/149#pullrequestreview-433547405) and the discussion on the forum: https://forums.ufscommunity.org/threads/strength-second-order-damping-top-layers

**Comment by climbfuji** [comment]
> This PR is to branch release/public-v1 and the PR discussion says that the branch for the UFS MRW App v1.1 release will be created from this branch. This is inconsistent with my understanding of the plan laid out by Arun in the 6/16 release meeting. I thought a release branch for the v1.1 release would be created from release/public-v1. I am not defending any particular solution, I just think this is inconsistent with my take away from the discussion and needs clarification. Can @arunchawla-NOAA clarify this and/or do we need more discussion?
> 
> On the topic of namelist changes, these look okay to me. They were proposed after a conversation with Fanglin as discussed in https://forums.ufscommunity.org/threads/strength-second-order-damping-top-layers.

Note that except for four lines of changes (the changes to the namelists), all other changes including those in ccpp-framework and the associated submodule pointer updates for ccpp-physics and fv3atm, are bugfixes for *existing code* in public/release-v1. These *must* go into these branches for any subsequent release, whether v1.0.1 (which I doubt will happen) or v1.1.0. I will rename the PRs for ccpp-framework and fv3atm accordingly, but keep the v1.1 reference in the title for ufs-weather-model because of the damping changes.

**Comment by climbfuji** [comment]
> > This PR is to branch release/public-v1 and the PR discussion says that the branch for the UFS MRW App v1.1 release will be created from this branch. This is inconsistent with my understanding of the plan laid out by Arun in the 6/16 release meeting. I thought a release branch for the v1.1 release would be created from release/public-v1. I am not defending any particular solution, I just think this is inconsistent with my take away from the discussion and needs clarification. Can @arunchawla-NOAA clarify this and/or do we need more discussion?
> > On the topic of namelist changes, these look okay to me. They were proposed after a conversation with Fanglin as discussed in https://forums.ufscommunity.org/threads/strength-second-order-damping-top-layers.
> 
> Note that except for four lines of changes (the changes to the namelists), all other changes including those in ccpp-framework and the associated submodule pointer updates for ccpp-physics and fv3atm, are bugfixes for _existing code_ in public/release-v1. These _must_ go into these branches for any subsequent release, whether v1.0.1 (which I doubt will happen) or v1.1.0. I will rename the PRs for ccpp-framework and fv3atm accordingly, but keep the v1.1 reference in the title for ufs-weather-model because of the damping changes.

FWIW, in an email conversation held earlier (March 25 this year) about the release strategy for the v1.1 release, it was decided to release v1.1 from release/public-v1. I am only posting an extract because I don't want to expose all names and email addresses here:

> From @junwang-noaa:
> I agree with Dom. For the versioning at EMC, we have hot fixes(z) in version number x.y.z and features upgrade (y) in x.y.z all go to release branch. Only the release upgrade (x) will be from the develop branch. E.g. GFSv15 and GFSv16 are from develop branch, GFS.v15.2.3 are from GFSv15 branch. So in my mind  the feature upgrade(python 3) should go to release branch. 

This email conversation was held between the MRW release managers and others, and based on the email the first round of Python 3 updates were merged into the release/public-v1 (v4 for CCPP) branches. I asked @arunchawla-NOAA for confirmation if we are sticking to the previous agreement or not.

**Comment by ligiabernardet** [comment]
Thank you @climbfuji  from seeking this clarification. I believe there is indeed an inconsistency between this plan charted in March and what I understood from the 6/15 meeting with release leads @arunchawla-NOAA @mvertens @nance. It will be good to get all in the same page. (cc @ceceliadid)

**Comment by arunchawla-NOAA** [comment]
The plan was to release v 1.1 from the release -v1 branch. But it was
discussed that we want the documentation to be at the top of the branch as
opposed to a particular tag of the branch. The reason for that was that
minor documentation updates require a tedious resetting of tags if the
documentation was set to a specific  tag. If the documentation was set to
the top of the branch then that would not be an issue. However, this
strategy does require us to have a separate branch if the documentation is
changing. So for minor hot fixes we would not create new branches as that
would not change documentation, but for mid level changes (aka v 1.1) we
would need to create a new branch so that documentation for v1.1 can be
seperate from that of v1. This does create additional branches but
hopefully not too many. And for that reason we would create a v 1.1 branch
from a v1 branch.
---------------------------------------------------------------
Arun Chawla
Chief
Engineering & Implementation Branch
Room 2083
National Center for Weather & Climate Prediction
5830 University Research Court
College Park, MD 20740
Phone: 301-683-3740
Mobile: 240-564-5675
     Fax: 301-683-3703
------------------------------------------------------------

**Comment by climbfuji** [comment]
> The plan was to release v 1.1 from the release -v1 branch. But it was discussed that we want the documentation to be at the top of the branch as opposed to a particular tag of the branch. The reason for that was that minor documentation updates require a tedious resetting of tags if the documentation was set to a specific tag. If the documentation was set to the top of the branch then that would not be an issue. However, this strategy does require us to have a separate branch if the documentation is changing. So for minor hot fixes we would not create new branches as that would not change documentation, but for mid level changes (aka v 1.1) we would need to create a new branch so that documentation for v1.1 can be seperate from that of v1. This does create additional branches but hopefully not too many. And for that reason we would create a v 1.1 branch from a v1 branch. --------------------------------------------------------------- Arun Chawla Chief Engineering & Implementation Branch Room 2083 National Center for Weather & Climate Prediction 5830 University Research Court College Park, MD 20740 Phone: 301-683-3740 Mobile: 240-564-5675 Fax: 301-683-3703 ------------------------------------------------------------
> […](#)

Thanks for the explanation, Arun. There must be a better way to do this. This just creates an unnecessary amount of work and goes against the EMC branching strategy for the GFS releases (as Jun alluded to in her email). Maybe the documentation needs to sit in a separate repository like we do it for CCPP, and that one gets pulled in as a submodule. Then all it takes to manage this is updating the submodule pointer for the documentation repo inside the ufs-mrweather-app (similar for ufs-weather-model). We could easily make this change in the release/public-v1 branch.

**Comment by arunchawla-NOAA** [comment]
ok I will set up a meeting to discuss this
---------------------------------------------------------------
Arun Chawla
Chief
Engineering & Implementation Branch
Room 2083
National Center for Weather & Climate Prediction
5830 University Research Court
College Park, MD 20740
Phone: 301-683-3740
Mobile: 240-564-5675
     Fax: 301-683-3703
------------------------------------------------------------

**Comment by climbfuji** [comment]
Back to the actual contents of this PR. In addition to the regression tests on hera and cheyenne, the code was tested to compile with `compile_cmake.sh` and `build.sh` on macOS with gcc+gfortran and clang+gfortran, and with Python 2.7 and Python 3.7.

**Comment by ligiabernardet** [comment]
One important consideration to decide this branching strategy is whether we
need to continue supporting the v1.0 release or whether that support will
end. And in more general terms, how many past releases, and which, we need
to support.

**Comment by climbfuji** [comment]
> One important consideration to decide this branching strategy is whether we need to continue supporting the v1.0 release or whether that support will end. And in more general terms, how many past releases, and which, we need to support.
> […](#)

I thought about this, and it actually is not an important point. I hope to share a slide or two later today to visualize this.

**Comment by climbfuji** [comment]
@ligiabernardet @junwang-noaa according to today's meeting, this PR can no go in - please approve and merge. Thanks!

File: parm/ccpp_v16beta_c96.nml.IN (hunk 1 above)
Comment (junwang-noaa) [reply parm/ccpp_v16beta_c96.nml.IN:52]: Just let you know that current GFSv16 forecast is using the updated coefficients for L127 levels, but GDAS forecast is using previous version.

**Review by junwang-noaa — APPROVED** [review]

**Review by ligiabernardet — APPROVED** [review]

**Pull request merged**